src/main/kotlin/org/rust/lang/refactoring/RsImportOptimizer.kt

```diff
@@ -72,18 +71,63 @@ class RsImportOptimizer : ImportOptimizer {
             val first = mod.childrenOfType<RsElement>()
                 .firstOrNull { it.textOffset >= offset && it !is RsExternCrateItem && it !is RsAttr } ?: return
             val psiFactory = RsPsiFactory(mod.project)
-            val sortedUses = uses
-                .sortedBy { it.useSpeck?.pathText }
-                .mapNotNull { it.copy() as? RsUseItem }
-            sortedUses
-                .mapNotNull { it.useSpeck }
-                .forEach { optimizeUseSpeck(psiFactory, it) }
-            for (importPath in sortedUses) {
-                mod.addBefore(importPath, first)
+            val sortedUsesGroups = uses
+                .map { UseItemWrapper(it) }
+                .groupBy({ it.packageGroupLevel }, { it })
+                .map { (groupLevel, useItemWrapper) ->
+                    groupLevel to useItemWrapper
+                        .sortedBy { it.useSpeckText }
+                        .mapNotNull { it.useItem.copy() as? RsUseItem }
+                }
+                .filter { it.second.isNotEmpty() }
+                .sortedBy { it.first }
+            sortedUsesGroups.forEach { sortedUses ->
+                sortedUses.second
+                    .mapNotNull { it.useSpeck }
+                    .forEach { optimizeUseSpeck(psiFactory, it) }
+            }
+
+            for ((_, sortedUses) in sortedUsesGroups) {
+                val lastUseItem = sortedUses.map { importPath ->
+                    val addedUseItem = mod.addBefore(importPath, first)
+                    mod.addAfter(psiFactory.createNewline(), addedUseItem)
```

Comment:
Interesting... it doesn't affect next item in the same group but without it, group separator group doesn't appear in the editor
But looks like we can drop it and insert two newlines after each group to have the same effect

Reply:
Yes, this is true.
However, I find the current version more logical. 
You do not think so, do you?

Reply:
I don't insist, it's a matter of taste